doc/good-practices.mld

```diff
@@ -0,0 +1,402 @@
+                      (create_record ~loc:ptype_loc fields, ptype_name)
+                  | { ptype_name; ptype_loc; _ } ->
+                      ( pexp_extension ~loc
+                        @@ Location.error_extensionf ~loc:ptype_loc
+                             "Cannot derive accessors for non record type %s"
+                             ptype_name.txt,
+                        ptype_name )
+                in
+                [
+                  pstr_value ~loc Nonrecursive
+                    [
+                      {
+                        pvb_pat = ppat_var ~loc { txt = "default_" ^ name.txt; loc };
+                        pvb_expr = e;
+                        pvb_attributes = [];
+                        pvb_loc = loc;
+                      };
+                    ];
+                ])
+            |> List.concat
+]}
+
+{2 In case of panic}
+
+In some rare cases, it might happen that a whole file rewriter is not able to
+output a meaningful AST. In this case, they might be tempted to raise a located
+error: an exception that includes the location of the error. Moreover, this has
+historically been what was suggested to do by ppxlib examples, but is now
+discouraged in most of the cases, as it prevents Merlin features to work well.
+
+If such an exception is uncaught, the ppx driver will return with an error code
+and the exception will be pretty-printed, including the location (that's the
+case when the driver is called by dune). When the driver is spawned with the
+[-embed-errors] or [-as-ppx] flags (that's the case when the driver is called by
+merlin), the driver will look for located error. If it catches one, it will stop
+its chain of rewriting at this point, and output an AST consisting of the
+located error followed by the last valid AST: the one passed to the raising
+rewriter.
+
+Even more in context-free rewriters, raising should be avoided, in favour of
+outputting a single error node when a finer grained reporting is not needed or
+possible. As the whole context-free rewriting is done in one traverse of the
+AST, a single raise will cancel both the context-free pass and upcoming
+rewriters, and the AST prior to the context-free pass will be outputted together
+with the error.
+
+The function provided by the API to raise located errors is
+{{!Ppxlib.Location.raise_errorf}[raise_errorf]}.
+
+{2 Migrating from raising to embedding errors}
+
+Lots of ppx-es exclusively use {{!Ppxlib.Location.raise_errorf}[raise_errorf]}
+to report errors, instead of the more merlin friendly way consisting of
+embedding errors in the AST, as described in this section.
+
+If you want to migrate such a codebase to the embedding approach, here are a few
+recipes to do that. Indeed, it might not be completely trivial, as raising can
+be done anywhere in the code, including in places where "embedding" would not
+make sense. What you can do is to turn your internal raising functions to
+function returning a [result] type.
```

Comment:
```suggestion
If you want to migrate such a codebase to the embedding approach, the rest of this section will present few
recipes to do that. It might not be completely trivial, as raising can
be done anywhere in the code, including in places where "embedding" would not
make sense. The first thing you can do is to turn your internal raising functions to
function returning a [result] type.
```